src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt

```diff
@@ -0,0 +1,389 @@
+            }
+        }
+    }
+}
+
+private class ChangeSignatureDialog(project: Project, descriptor: SignatureDescriptor) :
+    ChangeSignatureDialogBase<SignatureParameter,
+        RsFunction,
+        String,
+        SignatureDescriptor,
+        ModelItem,
+        TableModel
+        >(project, descriptor, false, descriptor.method) {
+    private var isValid: Boolean = true
+
+    private val config: RsChangeFunctionSignatureConfig
+        get() = myMethod.config
+
+    private var visibilityComboBox: VisibilityComboBox? = null
+
+    override fun getFileType(): LanguageFileType = RsFileType
+
+    override fun placeReturnTypeBeforeName(): Boolean = false
+
+    override fun createNorthPanel(): JComponent? {
+        val panel = super.createNorthPanel() ?: return null
+        // Make all two (or three) elements the same size
+        myNameField.setPreferredWidth(-1)
+        myReturnTypeField.setPreferredWidth(-1)
+
+        if (config.allowsVisibilityChange) {
+            val visibilityPanel = JPanel(BorderLayout(0, 2))
+            val visibilityLabel = JLabel("Visibility:")
+            visibilityPanel.add(visibilityLabel, BorderLayout.NORTH)
+
+            val visibility = VisibilityComboBox(project, config.visibility) { updateSignature() }
+            visibilityLabel.labelFor = visibility.component
+            visibilityPanel.add(visibility.component, BorderLayout.SOUTH)
+            visibilityComboBox = visibility
+
+            // Place visibility before function name and return type
+            val layout = panel.layout as GridBagLayout
+            val nameConstraints = layout.getConstraints(myNamePanel).clone() as GridBagConstraints
+            nameConstraints.gridx = 1
+            layout.setConstraints(myNamePanel, nameConstraints)
+
+            val myReturnTypePanel = myReturnTypeField.parent
+            val returnTypeConstraints = layout.getConstraints(myReturnTypePanel).clone() as GridBagConstraints
+            returnTypeConstraints.gridx = 2
+            layout.setConstraints(myReturnTypePanel, returnTypeConstraints)
+
+            val gbc = GridBagConstraints(0, 0, 1, 1, 1.0, 1.0,
+                GridBagConstraints.WEST,
+                GridBagConstraints.HORIZONTAL,
+                Insets(0, 0, 0, 0),
+                0, 0)
+            panel.add(visibilityPanel, gbc)
+        }
+        return panel
+    }
+
+    override fun createSouthAdditionalPanel(): JPanel {
+        // TODO: only in edition 2018?
+        val asyncBox = CheckBox("Async", config.isAsync)
+        asyncBox.addChangeListener {
+            config.isAsync = asyncBox.isSelected
+            updateSignature()
+        }
+        val unsafeBox = CheckBox("Unsafe", config.isUnsafe)
+        unsafeBox.addChangeListener {
+            config.isUnsafe = unsafeBox.isSelected
+            updateSignature()
+        }
+
+        return panel {
+            row {
+                asyncBox()
+                unsafeBox()
+            }
```

Comment:
We should find a way to center checkboxes vertically:

![image](https://user-images.githubusercontent.com/6505554/105079119-08923980-5aa0-11eb-81f1-46b2f6bdde50.png)

And also ideally add some padding between checkboxes and help icon

Reply:
I can create a gap with `.withLargeLeftGap()`, but I can't seem to figure out the vertical alignment. Any hints? There doesn't seem to be a lot of (documented) layout API available for the Kotlin UI.

We can also put them in a column:
![image](https://user-images.githubusercontent.com/4539057/105089735-cb2cac80-5a9d-11eb-8cfd-78b7539bd91f.png)

Reply:
Please try something like:
```kotlin
        return JPanel().apply {
            layout = MigLayout("align center center, insets 0 ${JBUI.scale(20)} 0 0")
            add(asyncBox)
            add(unsafeBox)
        }
```

Reply:
Oh, right, I don't have to use the UI DSL. Thanks for `MigLayout`, I didn't know about that.